### DISCUSSION TITLE
How to embed videos (`.mp4`, `.gif`, etc.)?

### DISCUSSION THREAD

**Discussion by UltraInstinct05** [squidfunk/mkdocs-material#3984] [General]
Has anyone ever embedded videos in Mkdocs for material to demonstrate particular concepts ? Any stuff you could recommend in terms of what software you use to make the videos/gifs and perhaps making to look good overall in light/dark mode etc.

**Comment by facelessuser** [comment]
Those are all good ways. It depends on what you are trying to do. Often, a simple GIF has worked for me in the past. I've used [LICEcap](https://www.cockos.com/licecap/) which works on macOS and windows. It's pretty basic but gets the job done. Your needs may be more complex than what I've generally needed though.

There are ways to potentially include dark and light variants of videos and only show one specific one during light or dark mode, but I honestly have never gone down that road. I usually just pick a scenario to record in and that is what I use in both light and dark mode.

Others may have solutions that may fit your needs more 🤷🏻 .

**Reply by UltraInstinct05** [reply]
Thanks for the inputs. This is just more of a general discussion. Often when you have such discussions, you discover great tools & techniques others use, but you may never have known.

**Reply by squidfunk** [reply]
We use [Streamable](https://streamable.com/) for the [Insiders demo video](https://squidfunk.github.io/mkdocs-material/insiders/). You can embed it via an `iframe`, i.e. something like:

``` html
<iframe src="https://streamable.com/e/yslhdu" allowfullscreen></iframe>
```

**Reply by UltraInstinct05** [reply]
I thought that particular page https://squidfunk.github.io/mkdocs-material/insiders/ always had an empty space. Never knew it was a video, because it never loaded for me on both Firefox & Chrome whenever I used the docs. I have a pretty decent internet connection speed, though.

![image](https://user-images.githubusercontent.com/34304644/171987643-54fe5128-a799-4f3d-ad90-f0b1dfb03f08.png)

![image](https://user-images.githubusercontent.com/34304644/171987656-c9c266ea-a840-4ebb-a16a-307ad6b89ec3.png)

**Reply by squidfunk** [reply]
Interesting, never seen that. It also works for me on slow connections. However, it might be related to a browser extension you're using. Try opening the page in incognito mode. EDIT: or Streamable just doesn't work in India 🤷‍♂️

**Reply by UltraInstinct05** [reply]
Your are right. Looks like it's blocked in India. https://www.reddit.com/r/india/comments/7wsa8s/streamable_not_working_in_india/

**Reply by squidfunk** [reply]
Meh.

**Comment by davideriboli** [comment]
If you can host your own video files, these lines work great.

```
<figure class="video_container">
  <video controls="true" allowfullscreen="true" poster="/path/to/poster.png">
    <source src="/path/to/video.mp4" type="video/mp4">
    <source src="/path/to/video.ogg" type="video/ogg">
    <source src="/path/to/video.webm" type="video/webm">
  </video>
</figure>
```

**Reply by gorger3** [reply]
Sorry, but this doesn't work. When built locally the page shows video block with the play button being inactive.
![image](https://github.com/squidfunk/mkdocs-material/assets/119797633/7c5fb978-b1eb-462d-966d-0de6f3c8dbab)

**Reply by kamilkrzyskow** [reply]
No need to message multiple times that it doesn't work.

What do you mean exactly by built locally?

What's the URL to the video in the embed?

Could be an issue with CORS if trying to open a video file from another server.

Perhaps you've used a relative path which isn't correct.

Please provide a minimal reproduction someone could try out and compare.

https://squidfunk.github.io/mkdocs-material/guides/creating-a-reproduction/

**Reply by gorger3** [reply]
Thank  you for reply.
I left two messages for two different solutions. I tried both on one page. I agree that it looks redundant.
By "built locally" I mean local server built by `mkdocs serve --dirtyreload` command.
The path is local. Video file is in the same folder as the page itself:
```
<video width="640"  controls>
    <source src="video-1.mp4" type="video/mp4">
</video>
```
File `video.html` in the same folder and with the same code opened in Chrome works well. So the path to the file is correct.

**Comment by HotCakeX** [comment]
This works perfectly

```html
<video controls>
<source src="VideoURL.mp4" type="video/mp4">
</video>
```

https://spynetgirl.github.io/Harden-Windows-Security-Module/Harden-Windows-Security-Module/#quick-demo

**Reply by gorger3** [reply]
As I wrote higher this, unforrtunately, doesn't work. When built locally the page shows video block with the play button being inactive.
![image](https://github.com/squidfunk/mkdocs-material/assets/119797633/0164eb14-ea2b-48dd-9ee7-d0a11e45c33a)

**Reply by HotCakeX** [reply]
Well it is working for me and you can clearly see it here: https://spynetgirl.github.io/Harden-Windows-Security-Module/Harden-Windows-Security-Module/#how-it-works

**Reply by alexvoss** [reply]
Hi, I think we would need more information so we have something to go by. Which browser are you are using? What have you done so far to rule out possible causes? Have you tried looking at the network traffic using the developer tools in the browser? Does the browser make a request for the video file? Anything reported on the server side?

**Reply by gorger3** [reply]
@alexvoss , thank you very much! Your advice helped me to find the cause. Chrome's devtools showed me that the browser "Failed to load resource: the server responded with a status of 404 (Not Found)"

Here is the directory tree:
- folder1
    - folder2
        - video-player.md
        - video.mp4
        - image.png

Results:

First try that failed.

Here I put the path to video file as I did it in Markdown syntax for images:

N | What I did in MkDocs' Markdown | What I saw in Chrome's HTML
--- | --- | ---
1 | `![alt text](image.png)` | `<img alt="alt text" src="../image.png">`
2 | `<source src="video.mp4" type="video/mp4">` | Failed to load resource: the server responded with a status of 404 (Not Found)<br>`http://127.0.0.1:8000/folder1/folder2/video-player/video.mp4`<br>Looking for a video file in a video-player folder that doesn't exist.

Second try that was successful.

Here I changed the path to video file according to HTML in Chrome:

N | What I did in MkDocs' Markdown | What I saw in Chrome's HTML
--- | --- | ---
1 | `![alt text](image.png)` | `<img alt="alt text" src="../image.png">`
2 | `<source src="../video.mp4" type="video/mp4">` | `<source src="../video.mp4" type="video/mp4">`<br>`http://127.0.0.1:8000/folder1/folder2/video.mp4`<br>Taking  the video file in the folder2 that contains the video file.

**Reply by blaylockbk** [reply]
Thanks for sharing this. I also had an issue linking a local video file. Had to do `../video.mp4` instead of `video.mp4` for whatever reason.

**Reply by alexvoss** [reply]
If you have [use_directory_urls](https://www.mkdocs.org/user-guide/configuration/#use_directory_urls) turned on in your `mkdocs.yml`, MkDocs will turn a file `video-player.md` into `video-player/index.html` so as to produce more user-friendly URLs ending in `video-player/`. This is the default. It means that your video now lives a level above the page that the user is viewing. That is why the video needs to be fetched from a level above.

**Reply by blaylockbk** [reply]
> If you have [use_directory_urls](https://www.mkdocs.org/user-guide/configuration/#use_directory_urls) turned on in your `mkdocs.yml`, MkDocs will turn a file `video-player.md` into `video-player/index.html` so as to produce more user-friendly URLs ending in `video-player/`. This is the default. It means that your video now lives a level above the page that the user is viewing. That is why the video needs to be fetched from a level above.

Thanks for explaining this reason!

**Reply by kolibril13** [reply]
my observation:
Version A) always works. Version B) ONLY works in index.md, not on other pages, even if they are not subchapters.
```
Version A)
<video controls src="../video.mp4" title="Title"></video>
Version B)
<video controls src="video.mp4" title="Title"></video>
```

**Comment by johnthagen** [comment]
It's unfortunately archived now, but I've used `mkdocs-video` to embed videos in a very simple way within Markdown. Perhaps someone will revive this project one day.

- https://github.com/soulless-viewer/mkdocs-video

```md
# Video example

Lorem ipsum dolor sit amet

![type:video](https://www.youtube.com/embed/LXb3EKWsInQ)
```

It supports relative paths to your own static file videos as well.
